File: python/packages/package-unit/src/package_unit/services/chunk_service.py
from __future__ import annotations

from typing import Any, Dict, List

from package_unit.services.embedding_service import EmbeddingService
from package_unit.services.summarization_service import SummarizationService
from package_unit.utils.token import estimate_token_count
from package_unit.utils.schema import make_schema
from package_unit.utils.sql import load_sql


class ChunkService:
    def __init__(self, db: Any) -> None:
        self.db = db

    def create_chunks(
        self,
        realm: str,
        tenant: str,
        team_id: str,
        unit_id: int,
        file_id: int,
        chunks: List[str],
        auth_headers: Dict[str, str],
    ) -> List[Dict[str, Any]]:
        if not chunks:
            return []
        created: List[Dict[str, Any]] = []
        for i, text in enumerate(chunks):
            summary = SummarizationService.summarize(text, auth_headers)
            embedding = EmbeddingService.get_embedding(summary, auth_headers)
            chunk_data = {
                "unit_id": unit_id,
                "file_id": file_id,
                "chunk_index": i,
                "text": text,
                "token_count": estimate_token_count(text),
                "summary": summary,
                "embedding": embedding,
                "metadata": {"chunk_size": len(text), "total_chunks": len(chunks)},
            }
            created_row = self.create_chunk(realm, tenant, team_id, chunk_data)
            created.append(created_row)
        return created

    def create_chunk(
        self,
        realm: str,
        tenant: str,
        team_id: str,
        chunk_data: Dict[str, Any],
    ) -> Dict[str, Any]:
        schema = make_schema(realm, tenant)
        sql = load_sql("chunks_insert.sql").format(schema=schema)
        values = [
            team_id,
            chunk_data["unit_id"],
            chunk_data["file_id"],
            chunk_data["chunk_index"],
            chunk_data["text"],
            chunk_data["token_count"],
            chunk_data["summary"],
            chunk_data["metadata"],
            chunk_data["embedding"],
        ]
        res = self.db.query(sql, values)
        return res.rows[0] if getattr(res, "rows", None) else res[0]

    def get_chunks_by_unit(
        self, realm: str, tenant: str, team_id: str, unit_id: int
    ) -> List[Dict[str, Any]]:
        schema = make_schema(realm, tenant)
        sql = load_sql("chunks_select_by_unit.sql").format(schema=schema)
        res = self.db.query(sql, [team_id, unit_id])
        return res.rows if getattr(res, "rows", None) is not None else res

File: node/packages/content/src/services/ChunkService.ts
import { ChunkData, CreateChunkRequest } from "../types";
import { EmbeddingService } from "./EmbeddingService";
import { SummarizationService } from "./SummarizationService";
import { estimateTokenCount } from "../utils/tokenEstimate";

export class ChunkService {
  constructor(private db: any) {}

  async createChunks(
    realm: string,
    tenant: string,
    teamId: string,
    unitId: number,
    fileId: number,
    chunks: string[],
    authHeaders: Record<string, string>
  ): Promise<ChunkData[]> {
    if (!chunks.length) return [];

    const createdChunks: ChunkData[] = [];

    for (let i = 0; i < chunks.length; i++) {
      const summary = await SummarizationService.summarize(chunks[i], authHeaders);
      const embedding = await EmbeddingService.getEmbedding(summary, authHeaders);

      const chunkData: CreateChunkRequest = {
        unit_id: unitId,
        file_id: fileId,
        chunk_index: i,
        text: chunks[i],
        token_count: estimateTokenCount(chunks[i]),
        summary,
        embedding,
        metadata: { chunk_size: chunks[i].length, total_chunks: chunks.length }
      };

      const created = await this.createChunk(realm, tenant, teamId, chunkData);
      createdChunks.push(created);
    }

    return createdChunks;
  }

  async createChunk(
    realm: string,
    tenant: string,
    teamId: string,
    chunkData: CreateChunkRequest
  ): Promise<ChunkData> {
    const query = `
      INSERT INTO ${realm}_${tenant}.chunks (
        team_id, unit_id, file_id, chunk_index, text,
        token_count, summary, metadata, embedding
      ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)
      RETURNING *`;

    const values = [
      teamId,
      chunkData.unit_id,
      chunkData.file_id,
      chunkData.chunk_index,
      chunkData.text,
      chunkData.token_count,
      chunkData.summary,
      JSON.stringify(chunkData.metadata),
      JSON.stringify(chunkData.embedding)
    ];

    const result = await this.db.query(query, values);
    return result.rows[0];
  }

  /**
   * Get all chunks for a specific unit
   */
  async getChunksByUnit(
    realm: string,
    tenant: string,
    teamId: string,
    unitId: number
  ): Promise<ChunkData[]> {
    const query = `
      SELECT * FROM ${realm}_${tenant}.chunks
      WHERE team_id = $1 AND unit_id = $2
      ORDER BY file_id, chunk_index;
    `;
    const result = await this.db.query(query, [teamId, unitId]);
    return result.rows;
  }
}
